cilium/CFP-30235-xds-adapter.md

```diff
@@ -0,0 +1,213 @@
+Cilium functionality. To start, it seems highly valuable to focus on the areas
+where existing xDS APIs overlap with existing Cilium capabilities.
+
+This shows how we could use existing xDS APIs to represent Cilium capabilities
+for ClusterIP routing:
+
+| Cilium Field | Origin xDS | Comments |
+| - | - | - |
+| Frontend.Address | `filter_chain_match.prefix_ranges` | |
+| Frontend.Port | `filter_chain_match.destination_port` | |
+| Frontend.Protocol | `cluster.metadata.protocol` | |
+| Backend[*].FEPortName | `cluster.metadata.port_name` | |
+| Backend[*].NodeName | ? | TBD - This is only used for logic to determine if an endpoint is local. Instead of sending this through xDS for every endpoint we could just look at the CIDR configured for each Cilium instance. |
+| Backend[*].Addr, Backend[*].Port | `lb_endpoints.endpoint.address.socket_address` | |
+| Backend[*].State | `lb_endpoints.health_status` | |
+| Backend[*].Preferred | N/A | Not used in Kubernetes |
+| Backend[*].Weight | `lb_endpoints.endpoint.load_balancing_weight` | |
+| Type | N/A | Out of scope - only focusing on ClusterIP Services |
+| ExternalTrafficPolicy, InternalTrafficPolicy | `cluster.load_balancing_policy` | Defined in the typed struct `type.googleapis.com/cilium_io_traffic_policy` |
+| NatPolicy | N/A | Not used in Kubernetes |
+| SessionAffinity | `tcpproxy.hash_policy` | |
+| SessionAffinityTimeout | ? | TBD - This represents the maximum session stickiness time. There is no natural place for this in existing xDS APIs. |
+| HealthCheckNodePort | N/A | Out of scope - only focusing on ClusterIP Services |
+| Name | `cluster.metadata.service_name` | |
+| Namespace | `cluster.metadata.service_namespace` | |
+
+### 3. Sample xDS Control Plane
+It would be useful for both the overall OSS experience and e2e testing to have
+an OSS xDS Control Plane that could serve as a reference xDS control plane
+implementation with Cilium. Of course there will likely be multiple options
+here, but it will be helpful to have at least a reference implementation for the
+purposes of e2e testing.
+
+### 4. Report Metrics via LRS
+One of the significant benefits of xDS is that it contains a load reporting
+mechanism. This can enable advanced routing capabilities at the xDS control
+plane layer, such as the topology aware routing example mentioned above. Within
+Cilium we'd likely want to report new and active connections to an xDS control
+plane via LRS.
+
+## Impacts / Key Questions
+
+### Key Question: Should we use a custom API on top of xDS-TP instead?
+Instead of building on top of existing xDS APIs, we could use a custom API on
+top of xDS-TP (xDS Transport Protocol). This would follow the pattern that
+Istio's Ztunnel project used when they developed
+[WDS](https://github.com/istio/ztunnel/blob/db0a74212c42c66b611a71a9613afb501074e257/proto/workload.proto).
+We could also just choose to send Cilium-specific protos over the wire.
```

Comment:
There's also the option of using existing Cilium agent API JSON types encoded using the proto Struct type, such that the adapter would be responsible for unmarshalling it to the Cilium API Go types (via JSON). Could you add this?

Reply:
Added, thanks!